Pun warnings (#765)
ghc-proposals/ghc-proposals: proposals/0000-pun-warnings.rst

============
Pun warnings
============

.. author:: Jaro Reinders, Artyom Kuznetsov
.. date-accepted:: 
.. ticket-url:: 
.. implemented:: 
.. highlight:: haskell
.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/765>`_.

> .. contents::

Reviewer: ```suggestion
.. sectnum::
.. contents::
```
Sections are not numbered

Author: https://github.com/ghc-proposals/ghc-proposals/issues/298

Reviewer: The proposal has "1 failing check" status (details of "All checks have failed") until you add `sectnum`

Author: Hmm... Then why was it removed from the template? And another random proposal I looked at also had no section numbers. But I guess I'll fix it if this proposal is necessary at all.

  Examples section).

* Introduce a new warning, ``-Wpuns`` and add it to ``-Weverything``. The warning is
  triggered by using an identifier that would be ambiguous or refer to another
  entity if Haskell had a single unified namespace. This includes syntactic puns
  like ``[]`` and ``()``.

Examples
========

Recall that ``-Wpun-bindings`` is triggered at definition sites that use punning,
and ``-Wpuns`` is triggered at use sites. To see what qualifies as punning, we
will look at the code that works today and analyze the breakage that would
occur if Haskell had a single unified namespace.

----------------------
``-Wpuns``, example #1
----------------------

::

  module A where { data A = T }
  module B where { data T = X }

  module C where

  import A
  import B

  f = T -- -Wpuns warning

If Haskell had a single unified namespace, referring to ``T`` would result in
ambiguity (is it ``A.T`` or ``B.T``?), thus this should trigger the warning.

Same happens if you use ``T`` in the export list:

::

  module C(T) where
      --  ^^^ -Wpuns warning
  import A
  import B

----------------------
``-Wpuns``, example #2
----------------------

::

  a = 15

>   f :: a -> a

Reviewer 2: This case is covered by the already implemented `-Wterm-variable-capture`. This doesn't preclude `-Wpuns` from triggering too, but it'd be useful to spell out the interaction.

I think `-Wterm-variable-capture` is a strict subset of `-Wpuns`.

Author: Yes, I think `-Wpuns` should have precedence over `-Wterm-variable-capture` if both are enabled.

means that punning is used and should trigger the warning.

On the contrary:

::

  a = 15

  f :: forall a. a -> a

Does not use punning because if Haskell had a single unified namespace, explicitly bound type variable ``a`` would shadow the top-level ``a``.

----------------------
``-Wpuns``, example #3
----------------------

::
  
  {-# LANGUAGE ScopedTypeVariabels #-}
  a = 15

  f :: forall a. a -> a
  f = \a -> (a :: a)
  --              ^ warning here

In all of the ``a`` uses except for the last one there is no punning, because if
Haskell had a single unified namespace, in the type signature, top-level ``a``
would be shadowed by explicitly bound type variable ``a``, and in the expression
``a`` variable bound in the lambda would shadow the type variable. In the very
last case, however, currently, the ``a`` would refer to the type variable, but if
Haskell had a single namespace it would refer to the term-level variable. Thus the
warning is triggered.

----------------------
``-Wpuns``, example #4
----------------------

::

  f :: [] a -- warning
  g :: [a]  -- warning
  g = []    -- warning
  h = [a]   -- warning
  x = [a,b] -- no warning

> Since ``-XListTupleTypeSyntax`` is enabled by default, all of the cases except the

Reviewer 2: The extension made it into GHC under the name `ListTuplePuns` rather than `ListTupleTypeSyntax`. There are a few places in this proposal than need to be updated accordingly.

write functions like this:

::

  sizeOf :: forall a -> Sized a => Int

  sizeOfInt = sizeOf Int

But because of punning, this can result in ambiguity:

::

  data T = T

  sizeOfT = sizeOf T -- is 'T' the data type or the data constructor?

We can't use the type of ``sizeOf`` to determine which ``T`` to use because of
*Lexical Scoping Principle*.

#281 tackles this issue by defaulting ``T`` to a data constructor in this case
(to keep compatibility with existing code) and introduces ``type`` syntactic marker.

However, thanks to *Syntactic Unification Principle* (adhered by `#281 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0281-visible-forall.rst>`_), if
the user chooses to not use punning, there is no need to use this syntactic
marker, resulting in less context-dependent and syntax cluttered code.

Unfortunately, GHC has no good support for such style of programming:

* There is no way for the user to know if some code uses punning without
  manually reviewing it, in other words, usage of punning is silent.

* There is no way to avoid punning if the user imports the code that makes the

>   use of it.

Reviewer 3: I believe this is a leftover from the original proposal that should be rephrased or removed now that we have `-XExplicitNamespaces` to do exactly this?

It would probably be good to mention somewhere in the proposal that `-XExplicitNamespaces` can be used to avoid triggering `-Wpuns` in these situations.

will look at the code that works today and analyze the breakage that would
occur if Haskell had a single unified namespace.

----------------------
``-Wpuns``, example #1
----------------------

::

  module A where { data A = T }
  module B where { data T = X }

  module C where

  import A
  import B

  f = T -- -Wpuns warning

If Haskell had a single unified namespace, referring to ``T`` would result in
ambiguity (is it ``A.T`` or ``B.T``?), thus this should trigger the warning.

Same happens if you use ``T`` in the export list:

::

  module C(T) where
      --  ^^^ -Wpuns warning
  import A
  import B

----------------------
``-Wpuns``, example #2
----------------------

::

  a = 15
  f :: a -> a

If Haskell had a single unified namespace, ``a`` instead of referring to
implicitly bound type-variable ``a`` would refer to ``a`` on the type-level. This
means that punning is used and should trigger the warning.

On the contrary:

::

  a = 15

  f :: forall a. a -> a

Does not use punning because if Haskell had a single unified namespace, explicitly bound type variable ``a`` would shadow the top-level ``a``.

> Note that this example is already covered by the ``-Wterm-variable-capture`` warning.

Reviewer 3: Further up the proposal states that `-Wterm-variable-capture` is a subset of `-Wpuns`. That seems at odds with this example being covered by `-Wterm-variable-capture` but not triggering `-Wpuns`.

When running this example in GHCi 9.14, it does not trigger `-Wterm-variable-capture` for me, surprisingly.

Reviewer 2: Here is a GHCi session, the warning seems to be firing as expected.

```
GHCi, version 9.14.1: https://www.haskell.org/ghc/  :? for help
ghci> :set -Wterm-variable-capture 
ghci> a = 15
ghci> f :: forall a. a -> a; f=f   -- no warning (as expected)
ghci> f :: a -> a; f=f             -- warning
<interactive>:4:1: warning: [GHC-54201] [-Wterm-variable-capture]
    The type variable ‘a’ is implicitly quantified,
    even though another variable of the same name is in scope:
      ‘a’ defined at <interactive>:2:1
    This is not compatible with the RequiredTypeArguments extension.
    Suggested fix: Consider renaming the type variable.
```

Reviewer 3: Ah, I see, I misread the text to mean that the second case is covered by `-Wterm-variable-capture`, thanks for clarifying.

I think ideally the text would state that the first case is what triggers the `-Wterm-variable-capture` warning.

  f = ()       -- warning
  g :: (a,b)   -- warning
  g = (c,d)    -- warning
  h :: (,) a b -- warning
  h = (,) c d  -- warning

Tuples in this case are very much the same as lists except they will emit a
warning in all cases.

Note that for both lists and tuples if ``-XListTuplePuns`` is disabled,
the type constructors will not be in scope anymore and no warnings will be
emitted.

------------------------------
``-Wpun-bindings``, example #1
------------------------------

This example shows the interaction with ``TypeAbstractions``.

::

  id :: t -> t
  id @a a = a

Here, when term-level ``a`` is bound it would conflict with the type level ``a`` if
Haskell had a single namespace, thus triggering the warning. This behavior is
similar to conflicting definition error for ``f b b = ...``:

::

  Test.hs:1:3: error:
      • Conflicting definitions for 'b'
        Bound at: Test.hs:1:3
                  Test.hs:1:5
      • In an equation for 'f'

On the contrary, the code below is fine, similarly to ``-Wpuns`` example #2, 
the ``a`` is shadowed instead:

::

  f :: t -> ()
  f @a = \a -> ()

Note how there is no conflicting definition and instead it is just shadowing 

> if both were term variables: ``f b = \b -> ...``. 

Reviewer 3: I think it would be clearer if the sentence used "would just be shadowing" instead of "is just shadowing", and if it mentioned that it (I assume) would trigger `-Wpuns` if the type variable `a` were used inside the lambda body.

------------------------------

::

  data T = T

If Haskell had a single unified namespace, type constructor ``T`` and data
constructor ``T`` would conflict, thus this should trigger the warning.

------------------------------
``-Wpun-bindings``, example #3
------------------------------

::

  data T = MkT
  data B = T | F

Even though type constructor ``T`` and data constructor ``T`` are defined in
different declarations, they would still cause a conflict, same as example #2.

------------------------------
``-Wpun-bindings``, example #4
------------------------------

::

  data J = Bool

This should not cause the warning because ``Bool`` defined here would not
conflict with ``Bool`` imported from ``Prelude``, this declaration is not rejected
by GHC:

::

  import Prelude (Bool)
  data Bool -- no conflict

------------------------------
``-Wpun-bindings``, example #5
------------------------------

This example shows the interaction with pattern signatures 
(part of ``ScopedTypeVariables``).

::

  f :: t -> t
  f @a = \(a :: a) -> a

This will not produce a punning warning, because there are
only two variables being bound, the first with ``@a`` and the second with 

> ``\(a :: ...) -> ...``.

Reviewer 3: This sentence is a bit confusing; only two variables being bound doesn't, by itself, seem like a good reason to not warn about them. I'm assuming the reason is that they live in different scopes.

Note that the ``:: a`` is just a use of the variable bound 
by the ``@a`` type abstraction.

> Renaming the second binding of ``a`` to ``x`` avoids shadowing:

Reviewer 3: Does this refer to actual shadowing that is happening or shadowing that would be happening if we had a unified namespace?

============
Pun warnings
============

.. author:: Jaro Reinders, Artyom Kuznetsov
.. date-accepted:: 
.. ticket-url:: 
.. implemented:: 
.. highlight:: haskell
.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/765>`_.
.. sectnum::
.. contents::

This proposal introduces ``-Wpun-uses`` and ``-Wpun-bindings``.

These changes should help the users write pun-free code to take advantage of
*Syntactic Unification Principle* described in `#378 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0378-dependent-type-design.rst>`_.

The warnings introduced in this proposal are intended

> help programmers enforce a pun-free style of programming in their Haskell

Reviewer 4: ```suggestion
to help programmers enforce a pun-free style of programming in their Haskell
```

Haskell makes heavy use of punning in its built-in syntax and common types:

::

  data [] a = [] | a : [a]
  data (a, b) = (a, b)
  data () = ()
  data Proxy a = Proxy
  newtype Identity a = Identity a
  newtype ReaderT r m a = ReaderT {runReaderT :: r -> m a}
  newtype ExceptT e m a = ExceptT (m (Either e a))

However, as Haskell's type system evolves, the distinction between types and
terms becomes blurry. For example, the ``-XDataKinds`` extension introduces the
``'`` syntax to select entities from the data namespace in a type-level context:

::

  r :: Rec Const '[ 'T ]   -- vinyl records
  r = Const :& RNil

Note that we had to qualify both the list syntax and the ``T`` data constructor
with a tick.

Syntactic Unification Principle (from `#378 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0378-dependent-type-design.rst>`_)
----------------------------------------------------------------------------------------------------------------------------------------------------

In the absence of punning, there is no difference between type-syntax and
term-syntax.

Syntactic Unification Principle implies that if the user is not using punning

> they can simply forget about the distinction between terms and types.

Reviewer 4: ```suggestion
they can simply forget about the distinction between terms and types for name resolution.
```

Lexical Scoping Principle (from `#378 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0378-dependent-type-design.rst>`_)
----------------------------------------------------------------------------------------------------------------------------------------------

For every occurrence of an identifier, it is possible to uniquely identify its
binding site, without involving the type system.

-----------------
Problem Statement
-----------------

> As we step towards Dependent Haskell (with `#378 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0378-dependent-type-design.rst>`_)

Reviewer 4: ```suggestion
As we step towards Dependent Haskell (with `#378 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0378-dependent-type-design.rst>`_
```

acceptance), the distinction between types and terms becomes blurrier and
blurrier and the need arises to use terms and types interchangeably. Indeed,
we can begin to see this need with ``RequiredTypeArguments`` which lets us 
write functions like this:

::

  sizeOf :: forall a -> Sized a => Int

  sizeOfInt = sizeOf Int

But because of punning, this can result in ambiguity:

::

  data T = T

  sizeOfT = sizeOf T -- is 'T' the data type or the data constructor?

We can't use the type of ``sizeOf`` to determine which ``T`` to use because of

> *Lexical Scoping Principle*.

Reviewer 4: ```suggestion
the *Lexical Scoping Principle*.
```

Proposal `#281 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0281-visible-forall.rst>`_ tackles this issue by defaulting ``T`` to a data constructor in this case

> (to keep compatibility with existing code) and introduces ``type`` syntactic marker.

Reviewer 4: ```suggestion
(to keep compatibility with existing code) and introduces the syntactic marker ``type``.
```

Author: Thank you!

the user chooses to not use punning, there is no need to use this syntactic
marker, resulting in less context-dependent and syntax cluttered code.

With the acceptance of the namespace-specified imports proposal (`#581 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0581-namespace-specified-imports.rst>`_), we now also have
the ability to specify which scope we want to import from: ::

  {-# LANGUAGE ExplicitNamespaces #-}

  import qualified Data.Proxy as T (type Proxy)   -- import only the Proxy type
  import qualified Data.Proxy as D (data Proxy)   -- import only the Proxy constructor

So it is always possible to avoid punning.

Unfortunately, GHC has no way for the user to know if some code uses punning
without manually reviewing it, in other words, usage of punning is silent.
This proposal addresses that problem: it lets users be confident their code
is pun-free by enabling this warning.

Beginner confusion
------------------

Even without Dependent Haskell, an argument can be made for pun-free code:

> punning can be a source of confusion for beginners. The difference between

Reviewer 5: not only for beginners

the terms namespace and the types namespace can be hard to understand at first,
especially when things like ``()`` or ``[a]`` are used (`as seen in this StackOverflow question <https://stackoverflow.com/questions/16892570/what-is-in-haskell-exactly>`_).

.. _Proposed Change Specification:

Proposed Change Specification
=============================

We propose to introduce two new warnings to GHC: ``-Wpun-uses`` and
``-Wpun-bindings`` and add them both to ``-Weverything``.

* ``-Wpun-uses`` warns the user about the usage of punning at use sites.

>   This is useful in several situations where ``-Wpun-bindings`` is does not warn:

Reviewer 3: ```suggestion
  This is useful in several situations where ``-Wpun-bindings`` does not warn:
```

Author: Oof, thanks for catching all these!

>   This is useful in several situations where ``-Wpun-bindings`` is does not warn:

Reviewer 6: At this point, `-Wpun-bindings` has not been introduced. Why though? I'd bring `-Wpun-bindings` before `-Wpun-uses` simply to avoid the confusion.

Reviewer 6: I also tend to think that the whole "this is useful because" para should come after the two bullets specifying the change.

>   * Multiple different imports might individually be pun-free, but when combined they can

Reviewer 5: We don't even warn about clashing names:

```
module Clashing where

import GHC.List
import Data.List
```

though at use case

```
*Clashing> null []

<interactive>:1:1: error:
    Ambiguous occurrence ‘null’
    It could refer to
       either ‘Data.List.null’,
              imported from ‘Data.List’ at Clashing.hs:4:1-16
              (and originally defined in ‘Data.Foldable’)
           or ‘GHC.List.null’,
              imported from ‘GHC.List’ at Clashing.hs:3:1-15
```

we report ambiguity error.

I think that if we go through a trouble of comparing imports into term and type namespaces for the purpose of puns, we also should report clashing names.

Author: Sorry, I didn't mean we will go through all names from imports. I just meant that this is not caught by `-Wpun-bindings` and that `-Wpun-uses` can catch these **if the punned names are used** (just like what we do with conflicting names as you describe).

Author: I will make this clearer in the proposal.

    give rise to puns, see example 1.

  * Some variables are bound implicitly, for example in type signatures without a ``forall``.
    These can be puns even though there is no binding site, see example 2.

  * Imported modules might simply not be pun-free.

  * The built-in list and tuple syntax uses punning, see wrinkle **W2** below and also example 4 and 5. 

* ``-Wpun-bindings`` warns the user about the introduction of punning at binding
  sites.

These warnings aim to help the user to avoid using punning in their codebase.

To determine whether some binding or use site takes advantage of punning we
ask the question: **"If Haskell had a single unified namespace, would that 
change the meaning of the program?"**. If the answer is yes, then the code
uses punning.

Wrinkles:

* **W1** Note that the hypothetical single-namespace version of Haskell would still have
  name shadowing, so the ``-Wpun-bindings`` warnings does not trigger if a name
  would merely be shadowed (i.e., redefined in a separate sub-scope).

* **W2** Furthermore, we include syntactic punning, for example using the ``[]`` or ``()`` syntax

>   triggers a warning from ``-Wpun-uses`` unless ``NoTupleListPuns`` is used. The proposal 

Reviewer 7: should be ``NoListTuplePuns`` I think?

different declarations, they would still cause a conflict, same as example #2.

------------------------------
``-Wpun-bindings``, example #4
------------------------------

::

  data J = Bool

This should not cause the warning because ``Bool`` defined here would not
conflict with ``Bool`` imported from ``Prelude``, this declaration is not rejected
by GHC:

::

  import Prelude (Bool)
  data Bool -- no conflict

------------------------------
``-Wpun-bindings``, example #5
------------------------------

This example shows the interaction with pattern signatures 
(part of ``ScopedTypeVariables``).

::

  f :: t -> t
  f @a = \(a :: a) -> a

Currently, pattern signatures, like ``a :: a`` in this case, may or may not
bind type variables depending on whether or not a variable with the same
name was already bound. In this case, the type variable ``a`` was
already bound by the ``@a`` type abstraction, so the pattern signature
is a use of the variable ``a`` and does not bind it as a fresh variable.

This will not produce a pun-use warning, because in the hypothetical pun-free
Haskell, this would be a would simply first bind ``a`` to be the type that 

> ``f`` operates on and subsequently shadows ``a`` to be the term argument to

Reviewer 3: ```suggestion
Haskell, this would simply first bind ``a`` to be the type that 
``f`` operates on and subsequently shadow ``a`` to be the term argument to
```

============
Pun warnings
============

.. author:: Jaro Reinders, Artyom Kuznetsov
.. date-accepted:: 
.. ticket-url:: 
.. implemented:: 
.. highlight:: haskell
.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/765>`_.
.. sectnum::
.. contents::

This proposal introduces ``-Wpun-uses`` and ``-Wpun-bindings``.

These changes should help the users write pun-free code to take advantage of
*Syntactic Unification Principle* described in `#378 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0378-dependent-type-design.rst>`_.

The warnings introduced in this proposal are intended
to help programmers enforce a pun-free style of programming in their Haskell
projects if they so desire. It does not claim that one style of programming
is strictly better than the other and acceptance or rejection of this proposal
should not mean that one style is preferred over the other by GHC.

Motivation
==========

----------
Background
----------

Before we move on to laying out the problem statement there are 3 concepts
that the reader is adviced to be familiar with: Punning, Syntactic

> Unification Principle and Lexical Scoping Principle.

Reviewer 8: ```suggestion
Before we move on to laying out the problem statement,
let us review three background concepts: Punning, the Syntactic
Unification Principle and the Lexical Scoping Principle.
```

In the absence of punning, there is no difference between type-syntax and
term-syntax.

Syntactic Unification Principle implies that if the user is not using punning
they can simply forget about the distinction between terms and types for name resolution.

Lexical Scoping Principle (from `#378 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0378-dependent-type-design.rst>`_)
----------------------------------------------------------------------------------------------------------------------------------------------

For every occurrence of an identifier, it is possible to uniquely identify its
binding site, without involving the type system.

-----------------
Problem Statement
-----------------

As we step towards Dependent Haskell (with `#378 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0378-dependent-type-design.rst>`_
acceptance), the distinction between types and terms becomes blurrier and
blurrier and the need arises to use terms and types interchangeably. Indeed,
we can begin to see this need with ``RequiredTypeArguments`` which lets us 

> write functions like this:

Reviewer 8: ```suggestion
As we step towards Dependent Haskell (with the acceptance of `#378 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0378-dependent-type-design.rst>`_),
the distinction between types and terms becomes blurrier and
blurrier, and the need arises to use terms and types interchangeably. Indeed,
we can begin to see this need with ``RequiredTypeArguments``, which lets us 
write functions like this:
```

::

  sizeOf :: forall a -> Sized a => Int

  sizeOfInt = sizeOf Int

But because of punning, this can result in ambiguity:

::

  data T = T

  sizeOfT = sizeOf T -- is 'T' the data type or the data constructor?

We can't use the type of ``sizeOf`` to determine which ``T`` to use because of
the *Lexical Scoping Principle*.

Proposal `#281 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0281-visible-forall.rst>`_ tackles this issue by defaulting ``T`` to a data constructor in this case
(to keep compatibility with existing code) and introduces the syntactic marker ``type``.

However, thanks to *Syntactic Unification Principle* (adhered by `#281 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0281-visible-forall.rst>`_), if
the user chooses to not use punning, there is no need to use this syntactic

> marker, resulting in less context-dependent and syntax cluttered code.

Reviewer 8: ```suggestion
marker, resulting in less context-dependent and syntax-cluttered code.
```

Even without Dependent Haskell, an argument can be made for pun-free code:
punning can be a source of confusion. It requires you to know whether something
is used in a term or a type context before you can mentally resolve a name.
Avoiding puns could slightly reduce the mental effort required to read code.

Newcomers are also not always aware that the same name may refer to two different 
entities, for example in `this StackOverflow question
<https://stackoverflow.com/questions/16892570/what-is-in-haskell-exactly>`_.

Proposed Change Specification
=============================

We propose to introduce two new warnings to GHC: ``-Wpun-uses`` and
``-Wpun-bindings`` and add them both to ``-Weverything``.

* ``-Wpun-uses`` warns the user about the usage of punning at use sites.

* ``-Wpun-bindings`` warns the user about the introduction of punning at binding
  sites.

These warnings aim to help the user to avoid using punning in their codebase.

To determine whether some binding or use site takes advantage of punning we
ask the question: **"If Haskell had a single unified namespace, would that 
change the meaning of the program?"**. If the answer is yes, then the code

> uses punning.

Reviewer 8: I wonder if we can capture the specification more crisply, rather than using a counterfactual?

It is not immediately obvious what it would mean for Haskell to have a single unified namespace. However it would (presumably) necessarily include enabling `NoListTuplePuns`, which means that code using lists/tuples only as data constructors would not change meaning, contrary to the interpretation of the counterfactual in W2 and the later examples.

Reviewer 9: > It is not immediately obvious what it would mean for Haskell to have a single unified namespace.

In particular, if my module doesn't use `DataKinds`, but I'm importing from a module that does. ( Presuming for the sake of illustration `Data.Bool` does, and using `ExplicitNamespaces`):

```
import qualified Data.Bool as B (type (Bool ..))
```

What names did I just import? Did I get both type `Bool` and type-of-type `Bool`? Did I get types `False, True`?

Author: Let me attempt a constructive definition:

One aspect of a Haskell program is scoping. Each AST node has a scope. For most nodes this will be equal to the scope of the parent node, but some nodes (like lambdas or let bindings) introduce a separate scope which include one or more new names, or it can shadow a name to have it refer to something else. 

Currently, every scope is split into two environments: one maps names to terms and the other maps names to types). Note that these include all kinds of names: top-level bindings, local bindings, constructors, type constructors, classes, etc. Some names are global and fixed, like `[]`,`()`, and `(,)`, but that just means they occur in every scope and there is no way to shadow them.

Unifying scopes is simple as long as there are no conflicting names in the environments. So ideally, we would warn whenever there is such a conflict. However, we also want to be able to implement this warning to have a running time that depends only on the size of the current module, so we cannot check this completely in the case of imports. A one line import may import an arbitrary number of names. So, we limit this check to two cases: 

1. An AST node (except for imports) introduces some number of new names. Here we can check whether it adds the same name to both the term and the type environments, in which case `-Wpun-bindings` triggers.
2. At a variable occurrence we can check whether its name is included in both the term and the type environments. In this case, `-Wpun-uses` triggers.

Note that (1) only warns when the same name is added to both environments simultaneously. It does not warn if a name is added to the term environment which already exists in the type namespace (or the other way around). We do not have to warn in this case because in a unified namespace the new name would simply shadow the existing name.

Oh, at this point I'm kind of stuck because of a problem with (2). As is, it would still warn even for names that would simply be shadowed if we had a unified namespace. I'll check what the proposed MR does.

Author: The MR simply tracks an additional unified environment. It is not yet clear to me how exactly it generates the warnings, but I can continue my constructive definition:

So we now track a third unified scope for every AST node. The contents of this scope can be defined in terms of how the term and type scopes change from one AST node to one of its children. If a name is added or redefined through shadowing in either the type or term scopes then the same is done for the unified scope. If such a change happens for the same name in both scopes then we mark it as conflicted in the unified scope (the change does not have to be the same in both scopes, for example a name could be added to the term scope and shadowed in the type scope). A conflicted name can still be shadowed by later bindings like any other name.

Using this definition of the unified scopes, we can define the `-Wpun-uses` warning as triggering when a variable is used which is marked as conflicted in the unified scope. The `-Wpun-bindings` can still work as I described in my previous comment.

Author: @AntC2

> What names did I just import? Did I get both type Bool and type-of-type Bool? Did I get types False, True?

Your syntax is incorrect. If you meant this:

```haskell
import qualified Data.Bool as B (type Bool (..))
```

Then you get the names `Bool`, `False`, and `True`. The `type` keyword is only used to disambiguate the name `Bool`, but in this case you're already also using the `(..)` notation which also already tells us that `Bool` must be a name from the type namespace, so the `type` keyword is redundant. This is true regardless of whether you are using `DataKinds` or not.

Author: @adamgundry I've written up my above sketch in the [detailed specification](https://github.com/ghc-proposals/ghc-proposals/blob/pun-warnings/proposals/0000-pun-warnings.rst#detailed-specification). Does that make sense to you and is it crisp enough?

Reviewer 8: Thanks @noughtmare! Your description does make sense to me. I've [suggested some editorial tweaks]( https://github.com/ghc-proposals/ghc-proposals/pull/765/changes#r3738449381), though alas I still feel vaguely dissatisfied that it ends up sketching an implementation without a great deal of precision. But in the absence of a good spec of the existing behaviour, perhaps this is the best we can feasibly do. And the spec can always be refined at the implementation stage. :smile:

Author: Thanks for your suggestions, I will merge them soon.

> it ends up sketching an implementation without a great deal of precision

Well, my intention was to give a precise *specification*. That's also why I started it with an explanation of a model of scopes.

Let me try to write the spec up as a Haskell program to make it even more precise.

Author: Alright, here is my specification:

```haskell
import Data.Map (Map, (!?))
import qualified Data.Map as Map

data Unique
instance Eq Unique

data Scope = MkScope
  { types :: Map String Unique
  , terms :: Map String Unique
  }

data Scope' = MkScope'
  { types' :: Map String Unique
  , terms' :: Map String Unique
  , unified' :: Map String (Maybe Unique) -- Nothing means conflict
  }

data Namespace = Ty | Tm

data AST scope = Node Unique scope [AST scope] | Var Namespace Unique String scope

unifiedScope :: Scope' -> Scope -> Scope'
unifiedScope (MkScope' tys tms uni) (MkScope tys' tms') = s' where
  tysd 
    | Map.isSubmapOfBy (\_ _ -> True) tys tys' = Map.differenceWith (\_ v -> Just v) tys tys'
    | otherwise = error "Improper type scope"
  tmsd
    | Map.isSubmapOfBy (\_ _ -> True) tms tms' = Map.differenceWith (\_ v -> Just v) tms tms'
    | otherwise = error "Improper term scope"
  uni' = Map.union uni (Map.unionWith (\_ _ -> Nothing {- -Wpun-binds warns here -}) (Map.map Just tysd) (Map.map Just tmsd))
  s' = MkScope' tys' tms' uni'

addUnifiedScope :: Scope' -> AST Scope -> AST Scope'
addUnifiedScope s (Var ns x name s') = Var ns x name (unifiedScope s s')
addUnifiedScope s (Node x s' xs) = Node x s'' (map (addUnifiedScope s'') xs) where
  s'' = unifiedScope s s'

data PunWarning
  = PunUseWarning Unique String
  | PunBindWarning Unique String

punWarnings :: AST Scope' -> [PunWarning]
punWarnings (Var ns x name (MkScope' tys tms uni)) = 
  case (env !? name, uni !? name) of
    (_, Just Nothing) -> [PunUseWarning x name] -- conflict
    (Just r, Just (Just r')) | r /= r' -> [PunUseWarning x name] -- shadowed in other namespace
    _ -> []
  where
    env = case ns of
      Tm -> tms
      Ty -> tys
punWarnings (Node x (MkScope' _ _ uni) xs) = pbws ++ concatMap punWarnings xs where
  pbws = [ PunBindWarning x name | (name, Nothing) <- Map.toList uni ]
```

Author: The code above has (at least) two issues:

1. It does account properly for the top level scope and top level bindings
2. It does not work for “non-shadowing” bindings, like the binding of type variables (without explicit for all) which only binds a new variable if a variable with that name does not already exist.

Author: <details>
<summary>Better prototype including the examples:</summary>

```haskell
module PunWarningsPrototype where

import Data.Map (Map, (!?))
import qualified Data.Map as Map
import Data.Maybe (catMaybes)

type Unique = String

data Scope = MkScope
  { types :: Map String Unique
  , terms :: Map String Unique
  , unified :: Map String (Maybe Unique) -- Nothing means conflict
  } deriving Show

-- we could add an explicit binding node
data AST scope = Node Unique NodeInfo scope [AST scope] deriving Show

-- if a binding cannot shadow then it just reuses the existing binding
-- this is used for implicitly quantified type variables (e.g. pun use example #2)
data Shadowing = CanShadow | CannotShadow deriving Show
data IsImport = YesImport | NotImport deriving Show
data Namespace = Ty | Tm deriving Show

data NodeInfo = Other | Bind Shadowing [(IsImport,Namespace,String,Unique)] | Var Namespace String
  deriving Show

var :: Unique -> Namespace -> String -> AST ()
var u ns name = Node u (Var ns name) () []

computeScopes :: Scope -> AST () -> AST Scope
computeScopes (MkScope tys tms uni) (Node x (Bind sh bs) () xs) = 
  let 
    tysd = Map.fromListWith (\_ _ -> error "name bound multiple times at once") [(name,u) | (_,Ty,name,u) <- bs]
    tmsd = Map.fromListWith (\_ _ -> error "name bound multiple times at once") [(name,u) | (_,Tm,name,u) <- bs]
    unid = Map.fromListWith (\_ _ -> Nothing) [(name,Just u) | (_,_,name,u) <- bs]
    f l r = case sh of CanShadow -> l; CannotShadow -> r
    s = MkScope
      (Map.unionWith f tysd tys)
      (Map.unionWith f tmsd tms)
      (Map.unionWith f unid uni)
  in Node x (Bind sh bs) s (map (computeScopes s) xs)
computeScopes parent (Node x ni () xs) = Node x ni parent (map (computeScopes parent) xs)

data PunWarning
  = PunUseWarning Unique String
  | PunBindWarning Unique String
  deriving Show

punBindWarnings :: AST a -> [PunWarning]
punBindWarnings (Node x (Bind _ bs) _ xs) =
  catMaybes (Map.elems (Map.fromListWithKey (\name _ _ -> Just (PunBindWarning x name)) [(name, Nothing) | (NotImport,_,name,_) <- bs]))
    ++ concatMap punBindWarnings xs
punBindWarnings (Node _ _ _ xs) = concatMap punBindWarnings xs

-- Add more built-in names if necessary
topLevelScope :: Scope
topLevelScope = MkScope 
  (Map.fromList [("[]", "Listtype"),("()","unittupletype"),("(,)","tupletype")])
  (Map.fromList [("[]", "emptylist"),("()","unittuplecon"),("(,)","tuplecon")])
  (Map.fromList [("[]", Just "emptylist"),("()", Just "unittuplecon"),("(,)", Just "tuplecon")])

punUseWarnings :: AST Scope -> [PunWarning]
punUseWarnings (Node x (Var ns name) (MkScope tys tms uni) []) = 
  case (env !? name, uni !? name) of
    (_, Just Nothing) -> [PunUseWarning x name] -- conflict
    (Just r, Just (Just r')) | r /= r' -> [PunUseWarning x name] -- shadowed in other namespace
      | otherwise -> []
    (Nothing, Just{}) -> error "punUseWarnings: improper scope"
    (Just{}, Nothing) -> error "punUseWarnings: improper scope"
    (Nothing, Nothing) -> []
  where
    env = case ns of
      Tm -> tms
      Ty -> tys
punUseWarnings (Node _ _ _ xs) = concatMap punUseWarnings xs where

punWarnings :: AST () -> [PunWarning]
punWarnings x = punBindWarnings x ++ punUseWarnings (computeScopes topLevelScope x)

imports :: [(Namespace, String, Unique)] -> [(IsImport, Namespace, String, Unique)]
imports = map (\(x,y,z) -> (YesImport,x,y,z))

notImports :: [(Namespace, String, Unique)] -> [(IsImport, Namespace, String, Unique)]
notImports = map (\(x,y,z) -> (NotImport,x,y,z))

punUseExample1 :: AST ()
punUseExample1 =
  Node "C" (Bind CanShadow (imports [(Ty,"A", "A.A"),(Tm,"T","A.T"),(Ty,"T","B.T"),(Tm,"X","B.X")])) () [
    Node "C.f" Other () [
      Node "C.f.body" (Var Tm "T") () []
    ]
  ]

-- >>> punWarnings punUseExample1
-- [PunUseWarning "C.f.body" "T"]

punUseExample2 :: AST ()
punUseExample2 =
  Node "Main" (Bind CanShadow (notImports [(Tm,"a","Main.a"),(Tm,"f","Main.f")])) () [
    Node "Main.f::" (Bind CannotShadow (notImports [(Ty,"a","Main.f::.a")])) () [
      Node "Main.f::._->" (Var Ty "a") () [],
      Node "Main.f::.->_" (Var Ty "a") () []
    ]
  ]

-- >>> punWarnings punUseExample2
-- [PunUseWarning "Main.f::._->" "a",PunUseWarning "Main.f::.->_" "a"]

punUseExample2fixed :: AST ()
punUseExample2fixed =
  Node "Main" (Bind CanShadow (notImports [(Tm,"a","Main.a"),(Tm,"f","Main.f")])) () [
    Node "Main.f::" (Bind CanShadow (notImports [(Ty,"a","Main.f::.a")])) () [
      Node "Main.f::._->" (Var Ty "a") () [],
      Node "Main.f::.->_" (Var Ty "a") () []
    ]
  ]

-- >>> punWarnings punUseExample2fixed
-- []

punUseExample3 :: AST ()
punUseExample3 =
  Node "Main" (Bind CanShadow (notImports [(Tm,"a","Main.a"),(Tm,"f","Main.f")])) () [
    Node "Main.f::" (Bind CanShadow (notImports [(Ty,"a","Main.f::.a")])) () [
      Node "Main.f::._->" (Var Ty "a") () [],
      Node "Main.f::.->_" (Var Ty "a") () []
    ],
    Node "Main.f=" (Bind CanShadow (notImports [(Ty,"a","Main.f::.a")])) () [
      Node "Main.f=.\\" (Bind CanShadow (notImports [(Tm,"a","Main.f=.\\")])) () [
        Node "Main.f=.\\._::" (Var Tm "a") () [],
        Node "Main.f=.\\.::_" (Var Ty "a") () []
      ]
    ]
  ]

-- >>> punWarnings punUseExample3
-- [PunUseWarning "Main.f=.\\.::_" "a"]

-- note that I skip 'g' here, because I assume '[a]' is already desugared to 
-- '[] a'
punUseExample4 :: AST ()
punUseExample4 =
  Node "Main" Other () [
    Node "Main.f::" Other () [
      Node "Main.f::._$" (Var Ty "[]") () [],
      Node "Main.f::.$_" (Var Ty "a") () []
    ],
    Node "Main.f=" (Var Tm "[]") () [],
    Node "Main.x=" Other () [],
    Node "Main.h::" Other () [
      Node "Main.h::._$" Other () [
        Node "Main.h::._$" (Var Ty "(,)") () [],
        Node "Main.h::.$_" (Var Ty "a") () []
      ],
      Node "Main.h::.$_" (Var Ty "b") () []
    ],
    Node "Main.h=" Other () [
      Node "Main.h=._$" Other () [
        Node "Main.h=._$" (Var Tm "(,)") () [],
        Node "Main.h=.$_" (Var Tm "a") () []
      ],
      Node "Main.h=.$_" (Var Tm "b") () []
    ]
  ]

-- >>> punWarnings punUseExample4
-- [PunUseWarning "Main.f::._$" "[]",PunUseWarning "Main.h::._$" "(,)"]

punBindExample1 :: AST ()
punBindExample1 =
  Node "Main" Other () [
    Node "id::" Other () [{- ... -}],
    Node "id=" (Bind CanShadow (notImports [(Ty,"a","id=0"),(Tm,"a","id=1")])) () [{- ... -}]
  ]

-- >>> punWarnings punBindExample1
-- [PunBindWarning "id=" "a"]

punBindExample1b :: AST ()
punBindExample1b =
  Node "Main" Other () [
    Node "id::" Other () [{- ... -}],
    Node "id=" (Bind CanShadow (notImports [(Ty,"a","id=0")])) () [
      Node "id=.\\" (Bind CanShadow (notImports [(Tm,"a","id=.\\")])) () [{- ... -}]
    ]
  ]

-- >>> punWarnings punBindExample1b
-- []

punBindExample2 :: AST ()
punBindExample2 =
  Node "Main" (Bind CanShadow (notImports [(Ty,"T","Main.T"),(Tm,"T","Main.T.T")])) () [
    -- ...
  ]

-- >>> punWarnings punBindExample2
-- [PunBindWarning "Main" "T"]

punBindExample3 :: AST ()
punBindExample3 =
  Node "Main" (Bind CanShadow $ notImports [
    (Ty,"T","Main.T"),
    (Tm,"MkT","Main.T.MkT"),
    (Ty,"B","Main.B"),
    (Tm,"T","Main.B.T"),
    (Ty,"T","Main.B.F")
    ]) () [
    -- ...
  ]

-- >>> punWarnings punBindExample3
-- [PunBindWarning "Main" "T"]

punBindExample4 :: AST ()
punBindExample4 =
  Node "Main" (Bind CanShadow $ imports [(Ty,"Bool","Prelude.Bool")] ++ notImports [
    (Ty,"J","Main.J"),
    (Tm,"Bool","Main.Bool")
    ]) () [
    -- ...
  ]

-- >>> punWarnings punBindExample4
-- []

-- This really does need a per-variable CanShadow, because the a in
-- 'f @_ = \(_ :: a) -> _' could be bound if it was not in scope.
-- For now I just hard-code that it does not shadow.

punBindExample5 :: AST ()
punBindExample5 =
  Node "Main" Other () [
    Node "Main.f=" (Bind CanShadow $ notImports [(Ty, "a", "Main.f=.a")]) () [
      Node "Main.f=.\\" (Bind CanShadow $ notImports [(Tm, "a", "Main.f=.\\")]) () [
        Node "Main.f=.\\._" (Var Tm "a") () []
      ]
    ]
  ]

-- >>> punWarnings punBindExample5
-- []
```
</details>

Author: @adamgundry I've added the executable specification to the proposal as an appendix.

If both ``-Wpun-uses`` and ``-Wterm-variable-capture`` are enabled the ``-Wpun-uses`` warnings
take precedence and the ``-Wterm-variable-capture`` warnings are suppressed.

--------------------------
``-Wpun-uses``, example #3
--------------------------

::
  
  {-# LANGUAGE ScopedTypeVariabels #-}
  a = 15

  f :: forall a. a -> a
  f = \a -> (a :: a)
  --              ^ warning here

In all of the ``a`` uses except for the last one there is no punning, because if
Haskell had a single unified namespace, in the type signature, top-level ``a``
would be shadowed by explicitly bound type variable ``a``, and in the expression
``a`` variable bound in the lambda would shadow the type variable. In the very
last case, however, currently, the ``a`` would refer to the type variable, but if
Haskell had a single namespace it would refer to the term-level variable. Thus the
warning is triggered.

Note that the ``-Wpun-bindings`` warning also triggers for the ``\a -> ...`` binder.

.. _example #4:

----------------------
``-Wpun-uses``, example #4
----------------------

::

  f :: [] a -- warning
  g :: [a]  -- warning
  g = []    -- warning
  h = [a]   -- warning

>   x = [a,b] -- no warning

Reviewer 8: Warning on term-level uses of list (or tuple) syntax seems somewhat counterproductive. After all, when `NoListTuplePuns` is used the term-level occurrences will become unambiguous again! Instead I'd suggest warning only on the type-level uses (i.e. those which will become errors under `NoListTuplePuns`). Indeed I would argue that doing so is closer to the counterfactual specification.

Author: Considering this again (with the help of my executable spec), I'm not sure if I agree.

The issue is that there would be no such thing as a “term-level” use if the namespaces were unified.

If we were to make a distinction like that then also -Wpun-uses example 1 would no longer trigger a warning.

Author: Ah, never mind. I was thinking that later imports would shadow earlier ones, but that's not how Haskell works, haha.

Reviewer 10: This is weird. That would mean you would not be able to write singleton lists at all with the warning. The warning should only trigger on uses of `[a]` that have type/kind `Type`. Because then it would mean the List type constructor which should be written as `List a`.

`[a]` should always mean a singleton list, no matter if on the term or type level and the warning should reflect that

From the examples above, only the first two should warn.

Author: @jvanbruegge Yep, I have already made that change as Adam suggested. I was now thinking if a similar line of thought would also lead to example 1 of the pun-uses to not trigger a warning, but I was mistaken.

unified environment is simply the union of the term and type environments.
At definition sites this is easy to check: every node that extends or shadows
names in a scope we check that it does not modify the same name in both the term
and type environments. If that does happen then ``-Wpun-bindings`` triggers.

At use sites, it is more difficult to check for punning, because, even if the
same name occurs in both the term and type environments, the name might simply
be shadowed if we had a unified namespace. To track this we introduce a third
unified environment. Whenever a name is modified (added or shadowed) in the term
or type environments, the same modification is applied to the unified
environment too. If the same name is modified in both the term and type
environments we mark the name as conflicted in the unified environment. 

Using this definition of the unified scopes, we can define the ``-Wpun-uses``
warning as triggering when a variable is used for which looking it up in the
unified environment yields a different result than looking it up normally. So,
either when that variable has been shadowed in a different namespace or when it

> is marked as conflicted.

Reviewer 8: ```suggestion
At definition sites, ``-Wpun-bindings`` triggers for a node that extends or
shadows names in a scope if it modifies the same name in both the term
and type environments.

At use sites, it is more difficult to check for punning, because even if the
same name occurs in both the term and type environments, the name might
have been shadowed if we had a unified namespace, making it unambiguous.
For example, `\ a -> (id :: forall a. a -> a)` does not trigger a warning on the uses
of `a`, because under a unified namespace, the `forall a` binder would shadow
the `\ a` binder.

To track this we introduce a third, unified environment, mapping names to terms
or types. Whenever a name is modified (added or shadowed) in the term
or type environments, the same modification is applied to the unified
environment too. (If the set of names in the term and type environments are disjoint, then the
unified environment is simply the union of the term and type environments.)
When the same name is modified in both the term and type
environments we mark the name as conflicted in the unified environment.
(This happens at definition sites for which `-Wpun-bindings` emits a warning,
but can also happen without a warning, e.g. where the same name is imported
as a type from one module and as a term from another.)

The ``-Wpun-uses`` warning triggers at a variable use site if looking up the variable in the
unified environment yields a different result from looking it up normally (in the
term or type environment, as appropriate). That is,
either the variable has been shadowed in a different namespace or it
has been marked as conflicted.
```
